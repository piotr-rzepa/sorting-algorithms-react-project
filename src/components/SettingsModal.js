/*
 * Zewnętrzny modal pozwalający użytkownikowi na konfigurację opcji do przeprowadzania testów
 */

/* eslint-disable react/prop-types */
import React, { useContext } from 'react';
import Modal from 'react-modal';
import { useDispatch, useSelector } from 'react-redux';
import {
	setSizeDataset,
	setSizeArray,
	setDataOrder,
	setSortingOrder,
	setArrayGain
} from '../actions/tests';
import { ThemeContext } from './ThemeContext';
import { styleModal, styleButtons, styleSearchBar } from '../styles/styles';

export default function SettingsModal({
	isOpen,
	onRequestClose,
	contentLabel
}) {
	const {
		datasetSize,
		arrayStartSize,
		dataOrder,
		sortingOrder,
		arrayEndSize
	} = useSelector((state) => state.tests);
	const dispatch = useDispatch();
	const darkTheme = useContext(ThemeContext);
	const [currentModalStyle, currentButtonStyle, currentInputStyle] = darkTheme
		? [styleModal.darkStyle, styleButtons.darkStyle, styleSearchBar.darkStyle]
		: [styleModal.lightStyle, styleButtons.lightStyle, styleSearchBar.lightStyle];
	return (
		<Modal
			isOpen={isOpen}
			onRequestClose={onRequestClose}
			contentLabel={contentLabel}
			closeTimeoutMS={200}
			className="modal"
			style={{ overlay: { background: currentModalStyle.backgroundColor } }}
		>
			<h2>Algorithm dataset settings</h2>
			<div className="numeric-settings">
				<div className="settings-title">
					<label htmlFor="dataset-length" name="dataset-length">
						Number of tests per array length:
					</label>
					<label htmlFor="starting-array-length" name="starting-array-length">
						Star array length:
					</label>
					<label htmlFor="array-gain" name="array-gain">
						End array length:
					</label>
				</div>
				<div className="settings-input">
					<input
						type="number"
						min="10"
						max="10000"
						step="10"
						value={datasetSize}
						onChange={(e) => dispatch(setSizeDataset(parseInt(e.target.value, 10)))}
						style={currentInputStyle}
						required
					></input>
					<input
						type="number"
						min="10"
						max="1000"
						step="10"
						value={arrayStartSize}
						onChange={(e) => dispatch(setSizeArray(parseInt(e.target.value, 10)))}
						style={currentInputStyle}
						required
					></input>

					<input
						type="number"
						min="100"
						max="10000"
						step="10"
						value={arrayEndSize}
						onChange={(e) => dispatch(setArrayGain(parseInt(e.target.value, 10)))}
						style={currentInputStyle}
						required
					></input>
				</div>
			</div>
			<div className="select-settings">
				<p>Select data order:</p>
				<div className="select-settings-single">
					<input
						type="radio"
						id="best"
						name="scenario"
						value="best"
						checked={dataOrder === 'best' ? true : false}
						onChange={() => dispatch(setDataOrder('best'))}
					></input>
					<label htmlFor="best"> Best scenario (data already sorted)</label>
				</div>
				<div className="select-settings-single">
					<input
						type="radio"
						id="worst"
						name="scenario"
						value="worst"
						checked={dataOrder === 'worst' ? true : false}
						onChange={() => dispatch(setDataOrder('worst'))}
					></input>
					<label htmlFor="worst"> Worst scenario (data sorted in reverse)</label>
				</div>
				<div className="select-settings-single">
					<input
						type="radio"
						id="random"
						name="scenario"
						value="random"
						checked={dataOrder === 'random' ? true : false}
						onChange={() => dispatch(setDataOrder('random'))}
					></input>
					<label htmlFor="random"> Random scenario (unfiorm distribution)</label>
				</div>
			</div>
			<div className="select-settings">
				<p>Select sorting order:</p>
				<div className="select-settings-single">
					<input
						type="radio"
						id="asc"
						name="sorting"
						value="asc"
						checked={sortingOrder === 'asc' ? true : false}
						onChange={() => dispatch(setSortingOrder('asc'))}
					></input>
					<label>Ascending order</label>
				</div>
				<div className="select-settings-single">
					<input
						type="radio"
						id="desc"
						name="sorting"
						value="desc"
						checked={sortingOrder === 'desc' ? true : false}
						onChange={() => dispatch(setSortingOrder('desc'))}
					></input>
					<label>Descending order</label>
				</div>
			</div>
			<button
				className="data-options-button--modal"
				type="button"
				onClick={() => onRequestClose()}
				style={currentButtonStyle}
			>
				Save and close
			</button>
		</Modal>
	);
}
