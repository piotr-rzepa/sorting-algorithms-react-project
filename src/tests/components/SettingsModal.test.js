import * as React from 'react';
import * as reactRedux from 'react-redux';
import { shallow } from 'enzyme';
import SettingsModal from '../../components/SettingsModal';
const data = [
	{
		id: 0,
		name: 'Bubble Sort',
		description:
			'Bubble sort, sometimes referred to as sinking sort, is a simple sorting algorithm',
		method: 'Exchanging',
		sortingTime: { best: 0, worst: 0, average: 10 }
	},
	{
		id: 1,
		name: 'Quick Sort',
		description: 'Quicksort is a divide-and-conquer algorithm.',
		method: 'Partitioning',
		sortingTime: { best: 0, worst: 0, average: 100 }
	}
];

jest.mock('../../selectors/fastslowestSelector', () => {
	jest.fn().mockImplementation(() => data);
});

describe('<Dashboard />', () => {
	const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
	const useContextMock = jest.spyOn(React, 'useContext');
	const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');
	beforeEach(() => {
		useSelectorMock.mockClear();
		useContextMock.mockClear();
		useDispatchMock.mockClear();
	});
	it('Should render component correctly', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		const wrapper = shallow(<SettingsModal />);
		expect(wrapper).toMatchSnapshot();
	});

	it('Should dispatch setSizeDataset correctly', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		wrapper.find('input').at(0).prop('onChange')({ target: { value: 10 } });
		expect(useDispatchMock).toHaveBeenCalled();
	});

	it('Should dispatch setArraySize correctly', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		wrapper.find('input').at(1).prop('onChange')({ target: { value: 10 } });
		expect(useDispatchMock).toHaveBeenCalled();
	});

	it('Should dispatch setArrayGain correctly', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		wrapper.find('input').at(2).prop('onChange')({ target: { value: 10 } });
		expect(useDispatchMock).toHaveBeenCalled();
	});

	it('Should dispatch setDataOrder correctly (best)', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		wrapper.find('input').at(3).prop('onChange')({ target: { value: 10 } });
		expect(useDispatchMock).toHaveBeenCalled();
	});

	it('Should change setDataOrder radiobutton (best) "checked" prop to true', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'best',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		expect(wrapper.find('input').at(3).prop('checked')).toBe(true);
	});

	it('Should change setDataOrder radiobutton (worst) "checked" prop to true', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'worst',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		expect(wrapper.find('input').at(4).prop('checked')).toBe(true);
	});

	it('Should change setDataOrder radiobutton (random) "checked" prop to true', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		expect(wrapper.find('input').at(5).prop('checked')).toBe(true);
	});

	it('Should dispatch setDataOrder correctly (worst)', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		wrapper.find('input').at(4).prop('onChange')({ target: { value: 10 } });
		expect(useDispatchMock).toHaveBeenCalled();
	});

	it('Should dispatch setDataOrder correctly (random)', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		wrapper.find('input').at(5).prop('onChange')({ target: { value: 10 } });
		expect(useDispatchMock).toHaveBeenCalled();
	});

	it('Should dispatch setSortingOrder correctly (asc)', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		wrapper.find('input').at(6).prop('onChange')({ target: { value: 10 } });
		expect(useDispatchMock).toHaveBeenCalled();
	});

	it('Should change setSortingOrder radiobutton (asc) "checked" prop to true', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'asc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		expect(wrapper.find('input').at(6).prop('checked')).toBe(true);
	});

	it('Should dispatch setSortingOrder correctly (desc)', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		wrapper.find('input').at(7).prop('onChange')({ target: { value: 10 } });
		expect(useDispatchMock).toHaveBeenCalled();
	});

	it('Should change setSortingOrder radiobutton (desc) "checked" prop to true', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const wrapper = shallow(<SettingsModal />);
		expect(wrapper.find('input').at(7).prop('checked')).toBe(true);
	});

	it('Should invoke closing modal correctly', () => {
		useSelectorMock.mockReturnValue({
			datasetSize: 100,
			arrayStartSize: 100,
			dataOrder: 'random',
			sortingOrder: 'desc',
			arrayEndSize: 1000
		});
		useContextMock.mockReturnValue(jest.fn());
		useDispatchMock.mockReturnValue((x) => x);
		const onRequestCloseMock = jest.fn(() => true);
		const wrapper = shallow(
			<SettingsModal onRequestClose={onRequestCloseMock} />
		);
		const result = wrapper.find('button').at(0).prop('onClick')();
		expect(result).toMatchSnapshot();
		expect(result).toBe(true);
		expect(result).toBeTruthy();
	});
});
