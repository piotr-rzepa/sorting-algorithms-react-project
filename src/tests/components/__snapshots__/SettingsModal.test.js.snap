// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> Should invoke closing modal correctly 1`] = `true`;

exports[`<Dashboard /> Should render component correctly 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentElement={[Function]}
  isOpen={false}
  overlayElement={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  preventScroll={false}
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
  style={
    Object {
      "overlay": Object {
        "background": "#949496",
      },
    }
  }
>
  <h2>
    Algorithm dataset settings
  </h2>
  <div
    className="numeric-settings"
  >
    <div
      className="settings-title"
    >
      <label
        htmlFor="dataset-length"
        name="dataset-length"
      >
        Number of tests per array length:
      </label>
      <label
        htmlFor="starting-array-length"
        name="starting-array-length"
      >
        Star array length:
      </label>
      <label
        htmlFor="array-gain"
        name="array-gain"
      >
        End array length:
      </label>
    </div>
    <div
      className="settings-input"
    >
      <input
        max="10000"
        min="10"
        onChange={[Function]}
        required={true}
        step="10"
        style={
          Object {
            "backgroundColor": "#606060",
            "color": "white",
          }
        }
        type="number"
        value={100}
      />
      <input
        max="1000"
        min="10"
        onChange={[Function]}
        required={true}
        step="10"
        style={
          Object {
            "backgroundColor": "#606060",
            "color": "white",
          }
        }
        type="number"
        value={100}
      />
      <input
        max="10000"
        min="100"
        onChange={[Function]}
        required={true}
        step="10"
        style={
          Object {
            "backgroundColor": "#606060",
            "color": "white",
          }
        }
        type="number"
        value={1000}
      />
    </div>
  </div>
  <div
    className="select-settings"
  >
    <p>
      Select data order:
    </p>
    <div
      className="select-settings-single"
    >
      <input
        checked={false}
        id="best"
        name="scenario"
        onChange={[Function]}
        type="radio"
        value="best"
      />
      <label
        htmlFor="best"
      >
         Best scenario (data already sorted)
      </label>
    </div>
    <div
      className="select-settings-single"
    >
      <input
        checked={false}
        id="worst"
        name="scenario"
        onChange={[Function]}
        type="radio"
        value="worst"
      />
      <label
        htmlFor="worst"
      >
         Worst scenario (data sorted in reverse)
      </label>
    </div>
    <div
      className="select-settings-single"
    >
      <input
        checked={true}
        id="random"
        name="scenario"
        onChange={[Function]}
        type="radio"
        value="random"
      />
      <label
        htmlFor="random"
      >
         Random scenario (unfiorm distribution)
      </label>
    </div>
  </div>
  <div
    className="select-settings"
  >
    <p>
      Select sorting order:
    </p>
    <div
      className="select-settings-single"
    >
      <input
        checked={false}
        id="asc"
        name="sorting"
        onChange={[Function]}
        type="radio"
        value="asc"
      />
      <label>
        Ascending order
      </label>
    </div>
    <div
      className="select-settings-single"
    >
      <input
        checked={true}
        id="desc"
        name="sorting"
        onChange={[Function]}
        type="radio"
        value="desc"
      />
      <label>
        Descending order
      </label>
    </div>
  </div>
  <button
    className="data-options-button--modal"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#949496",
      }
    }
    type="button"
  >
    Save and close
  </button>
</Modal>
`;
